### PR TITLE
Session Store: Main implementation

### PR DESCRIPTION
**Pull request by apourchet**

### CODE REVIEW

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/SessionMetadataDoc.java
Comment (agavra) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/SessionMetadataDoc.java:26]: I think it might make sense to share this and `WindowMetadataDoc` for now - including all the code in `MongoSessionTable` related to metadata management

File: kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveSessionParams.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveSessionParams.java:42]: Might have forgotten to mention this before --there is no "retain duplicates" for session stores. It's a special parameter of the WindowStore specifically. One could argue it's a bit of a hack, but basically the WindowStore gets reused by Kafka Streams for stream-stream joins, where it effectively treats the WindowStore as a regular KV-plus-timestamp store. But it has to be able to hold multple values for the same key-timestamp since you could have multiple events with the same key and timestamp in your join, thus Streams will set `retainDuplicates` to true for WindowStores that are to be used for stream-stream joins (as opposed to aggregations, which should have strict update semantics)

tl;dr we should remove all the retainDuplicates stuff for session stores

File: kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveSessionParams.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveSessionParams.java:109]: Just an aside: I think it'll be a lot harder for us to come up with a "smart" choice for session stores. We can leave the TODO of course, but this is probably a case where we want to set a reasonable default and then in practice work with our users to figure out a reasonable value

File: kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveStores.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveStores.java:67]: Is your IDE set to autoformat? I usually turn that off just because it can make an already-long PR even longer -- but it's not a big deal. To me, at least -- if you submit a PR to Kafka/Kafka Streams, some of the other committers _will_ complain about extraneous formatting changes. Even if they are an improvement lol. Just a heads up

File: kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveStores.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveStores.java:321]: This should be set to the value of `truncateChangelog` that's in the session params

I'm pretty sure this was an existing bug though, we should probably fix that in a separate PR. We might want to be careful to ensure that Metronome sets this correctly -- @agavra do they still want to truncate the changelog? I'm a little hesitant after it saved us hahaha. Should we wait to fix this bug?

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/MongoSessionFlushManager.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/db/MongoSessionFlushManager.java:40]: Is this class basically the same as the MongoWindowFlushManager except for the type of this table (and the partitioner)? Just wondering, did you look into whether it's possible to consolidate them by parametrizing somehow?

My main concern with all this is: how easy would it be to make a change in one class but not the other, and accidentally screw up one of the implementations. It's not the end of the world to have repeated code, but what's the risk of code which should be identical getting out of sync?

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/SessionDoc.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/SessionDoc.java:76]: QQ: are the getter names enforced by the bson format? If not, let's drop the `get` from this to be consistent with naming elsewhere in the code

(I know we're inconsistent in some of the other Mongo Doc classes right now, there was some confusion around what was required for bson objects. And clearly we still got some things wrong)

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/SessionDoc.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/SessionDoc.java:36]: nit: can we change the suffix from TS to MS? I know I myself used TS in the WindowDoc (and we should change that too) but it's pretty clearly a timestamp whereas MS at least tells us the units. And we tend to use `Ms` suffixes for similar variable names in Kafka Streams

File: kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java
Comment (agavra) [inline kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java:132]: can we test things outside the inactivity gap and also inside the inactivity gap but requiring the grace period (i.e. out of order events)?

File: kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java
Comment (agavra) [inline kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java:172]: 3s might be too fast for the janky CI machines -- consider bumping this up to something like 20s (since it's expected to pass fast, it's OK if the timeout is slow)

File: kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java
Comment (agavra) [inline kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java:185]: any reason not to just merge this with the first test? (e.g. does it use different functionality)

File: kafka-client/src/test/java/dev/responsive/kafka/internal/db/SessionKeySpecTest.java
Comment (agavra) [inline kafka-client/src/test/java/dev/responsive/kafka/internal/db/SessionKeySpecTest.java:35]: please use given,when,then and name `shouldXYZ` 😉 the one nit I'm passionate about haha

File: kafka-client/src/test/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTableTest.java
Comment (agavra) [inline kafka-client/src/test/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTableTest.java:65]: are there any meaningful negative tests we should test? (i.e. throwing exceptions)

also suggestion - we should consider splitting out: `shouldFindInsertedKeyWithValidStartTimestamp`, `shouldNotFindInsertedKeyWithInvalidStartTimestamp` and `shouldNotFindKeyThatWasNotInserted`

That way it's very clear what each of the tests are doing and when they fail the test name says it all

File: kafka-client/src/test/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTableTest.java
Comment (agavra) [inline kafka-client/src/test/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTableTest.java:154]: feel like we should test this with more than one result!

File: kafka-client/src/test/java/dev/responsive/kafka/testutils/IntegrationTestUtils.java
Comment (agavra) [inline kafka-client/src/test/java/dev/responsive/kafka/testutils/IntegrationTestUtils.java:341]: personal preference, but I prefer to have these in the test file directly so I don't need to read two files (especially when reviewing) to see what a test is doing especially since it's important to the output

**Review by agavra — COMMENTED** [review]
I haven't actually thought at all about the most important class yet (`MongoSessionTable`) but figured I'd get all the minor comments out of the way first. Most of them are about code duplication and seeing if we can reduce it a bit.

File: kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveStores.java
Comment (agavra) [reply kafka-client/src/main/java/dev/responsive/kafka/api/stores/ResponsiveStores.java:321]: I'm definitely supportive of not fixing it in this PR. I do think we should consider removing changelog truncation altogether.

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/MongoSessionFlushManager.java
Comment (apourchet) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/db/MongoSessionFlushManager.java:40]: You're right, it is almost entirely a copy paste of the MongoWindowFlushManager class, which itself is a relatively close replica of the MongoKVFlushManager. My plan was to refactor these 3 into 1 separately in another PR so as to keep this one entirely devoid of any changes impacting other stores (KV / Window).

File: kafka-client/src/test/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTableTest.java
Comment (apourchet) [reply kafka-client/src/test/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTableTest.java:72]: No negative test that I can think of right now, I wrote those tests as mirrors of the ones for the KV and Window stores but we might want to expand all of them in similar ways (to include negative tests as you mentioned).

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/SessionMetadataDoc.java
Comment (apourchet) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/SessionMetadataDoc.java:26]: I agree that this needs to be consolidated. I wrote this down and will address in a follow-up PR.

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java
Comment (ableegoldman) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java:168]: Actually yes, they are deleted as part of the regular session window aggregation. While session stores are similar to window stores but with an additional timestamp, how they are used (in the DSL operators) is quite different. 

Window stores are essentially simple read-update-write -- session stores are more like rangeScan-maybeMerge-maybeDelete

Deletes don't necessarily happen on every record but they are a foundation of the session store

File: kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java
Comment (ableegoldman) [reply kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java]: Just to build off of this -- given the startup time for our integration tests, it often makes sense to throw a bit of "good testing practice" out the window and try to do as much as possible within a single test. If it's possible to combine tests without sacrificing the coverage, we should do so.

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java:127]: This comment doesn't quite make sense since we switched from segment ids to start timestamps. I think you can just replace this line with `activeSegments.get(activeSegments.length() - 1)).segmentStartTimestamp

(Probably also makes sense to just inline `activeSegments.get(0).segmentStartTimestamp` and remove the `firstSegmentId` variable too)

Comment (ableegoldman) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java:127]: Can you do a quick check to see if this needs to be updated in the MongoWindowTable as well? This feels a bit familiar so I'm wondering if maybe this was just based on the older version of the window table code before your other PR changed it

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java:141]: nit: same here, should be segmentStartTimestamp -- I'm guessing this is just a holdover from before your last PR, so maybe just give that a quick pass to see what changed in MongoWindowTable during the segmentId --> segmentStartTimestamp update and now needs to be reflected here too?

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java:163]: I'm still a bit hesitant about whether we want to support it at all for window stores, but as noted already, we clearly need to for KV stores and session stores.

However I don't think there's any reasonable need to configure it per store. Maybe we can just throw in a global config for tombstone retention and call it a day? (Not in this PR of course)  cc @agavra

File: kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java
Comment (ableegoldman) [inline kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java:171]: I know you just copied this over so it's probably @agavra who I should aim this comment at, but -- just FYI, the `APPLICATION_SERVER_CONFIG` is really only for IQ and giving instances a name that the user can recognize/control. In other words, we definitely don't need to set it in any of our functional tests (IQ == Interactive Queries, a Streams feature for reading the state store contents from outside the app)

File: kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java
Comment (ableegoldman) [inline kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java:186]: Guessing we can remove this 😄

File: kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java
Comment (ableegoldman) [inline kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java:144]: I'm confused...the inactivity gap is 5000, right? So if a is at 0 and b is at 6000, then there's more than 5000ms between these two events, and therefore they should not be merged into the same session...right?

It's been a long time since I thought about session aggregations so it's completely possible I'm not looking at this the right way. But my understanding is that  events separated by more than the inactivity gap should stay separate, at least unless/until another event comes in between them and is within the inactivity gap of both. In that case you get could a and b merged with each other -- if, say, c had timestamp 3000 for example, we would see "a" and "b" merged into "abc". But I think (not 100% sure) that we'd never see just an "ab" output

File: kafka-client/src/main/java/dev/responsive/kafka/internal/stores/SessionOperationsImpl.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/stores/SessionOperationsImpl.java:244]: parameter name should be latestSessionStart

File: kafka-client/src/main/java/dev/responsive/kafka/internal/stores/SessionOperationsImpl.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/stores/SessionOperationsImpl.java:251]:
```suggestion
          && kv.key.sessionStartMs <= latestSessionEnd;
```

I'm wondering if we should have a separate integration test where we _don't_ set the CommitBuffer to flush on every record -- sadly the soak won't catch things like this since we don't do any correctness checking at the moment (of course eventually we hope to use your new store-comparison tool for that). It's still probably worth doing an integration test.

Maybe to save time we could knock out all three store types in the same integration test with less-frequent buffer flushing (rather than having an extra test for this in each store type's integration test class)? This would be followup work of course, just thinking out loud. For now, maybe just try running the test once with the buffer flush config commented out, to make sure the results line up (but merging it with the buffer flush config re-enabled)

File: kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java
Comment (ableegoldman) [inline kafka-client/src/test/java/dev/responsive/kafka/integration/ResponsiveSessionStoreIntegrationTest.java:230]: This is the config I was referring to above, just fyi -- worth running just once with this disabled

File: kafka-client/src/test/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTableTest.java
Comment (ableegoldman) [reply kafka-client/src/test/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTableTest.java:72]: Expanding the basic test coverage (ie unit tests) of all the stores has been on my TODO list for a while. It's hard to make time to prioritize writing tests for things that seem to be working when there is so much else to do -- but we don't want to wait until it's too late to catch bugs. 

Perhaps we could dedicate something like one day each month towards "bug bashing and bug prevention" -- including filling in holes in our test coverage. I think it would be pretty valuable. Even a half day a month could pay off if all of us do it together

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java:148]:
```suggestion
          "{}[{}] Creating segment with start timestamp {}",
```

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java:190]:
```suggestion
          "{}[{}] Expiring segment with start timestamp {}",
```

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java:372]:
```suggestion
      final long latestSessionStart
```

ditto elsewhere this appears

**Review by ableegoldman — APPROVED** [review]
A bunch of minor things (eg parameter name updates) but all in all this LGTM.

I don't think there was anything that would need another pass so feel free to merge when you've addressed all the comments (obviously excluding the ones about followup work outside of this PR). I'm also happy to take another look at anything you want me to do another pass over before merging, as usual just give me a ping if so

File: kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java
Comment (apourchet) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java:372]: I am actually using earliest / latest session end to filter the remote table and find the sessions.

File: kafka-client/src/main/java/dev/responsive/kafka/internal/stores/SessionOperationsImpl.java
Comment (apourchet) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/stores/SessionOperationsImpl.java:244]: Same as for the RemoteSessionTable interface, I only filter based on session end, so I pass in [earliest, latest] as a way to lower + upper bound the session end.
The latestSessionStart timestamp is only used as a last-minute filter on the results from filtering the session end.

File: kafka-client/src/main/java/dev/responsive/kafka/internal/stores/ResponsiveSessionStore.java
Comment (apourchet) [reply kafka-client/src/main/java/dev/responsive/kafka/internal/stores/ResponsiveSessionStore.java:138]: I think this is left-over from when I pasted the comments from the javadocs. They're not good comments to have in the first place so I will remove them.

**Pull request merged**
